## alphaPhysical

alphaPhysical is a tool for exploring the biophysical properties of protein structures
generated using Alphafold.

Given one or (preferably) more protein sequences in FASTA format, alphaPhysical iterates these through
a deep neural network trained on the latest genome and protein databases, then extracts
and plots various biophysical properties of the predicted, AMBER-relaxed structures.

alphaPhysical may facilitate hypothesis generation and testing regarding the function of novel
sequences within protein families. alphaPhysical is also intended for exploring the
biophysical properties of bespoke protein sequences.

alphaPhysical is currently in very early development and testing.


## Requirements

alphaPhysical will run on almost any system with at least 16GB available RAM and modest storage.
[WSL2](https://learn.microsoft.com/en-us/windows/wsl/install) is recommended for Windows users.

For structural predictions, [a capable GPU](https://www.nvidia.com/en-us/design-visualization/desktop-graphics/#nv-accordion-bd50fbb79d-item-7577526115) and several TB of high-speed storage are required.


## Installation

1. Install [Docker](https://www.docker.com/).

2. Clone this repository and cd into it.

```
git clone http://localhost:18068/steve/alphaPhysical.git && cd alphaPhysical
```

3. Build the docker image.

```
docker build -f Dockerfile -t alphaphysical .
```

4. Install [Alphafold and necessary databases](https://github.com/google-deepmind/alphafold).

5. Configure alphaPhysical.

Edit the following lines of the run_alphaPhysical.sh main runscript: 

    alphafold_path="/mnt/d/analyze/alphafold/run"		# your path where Alphafold was cloned in step 4.
    alphafold_db="/mnt/e/alphafold/database"		# your path where databases were downloaded in step 4.
    fasta_in="/mnt/d/analyze/alphafold/target/active"	# your path where FASTA sequences of interest are located.
    output="/mnt/d/analyze/alphafold/target"		# your path where all outputs are written.


## Usage

The recommended way to run alphaPhysical is to edit the main runscript as described above,
then invoke the main runscript as follows.


**Default run example:**

```
bash run_alphaPhysical.sh

```

It is also possible to configure an alphaPhysical run at the command line.
See `bash run_alphaPhysical.sh --help` for more details.


**Command line-configured run example:**

```
bash run_alphaPhysical.sh \
	--alphafold_path [str] \
	--alphafold_db [str] \
	--fasta_in [str] \
	--output [str]

```

Note that most protein predictions will take several hours.


## Output

A successful alphaPhysical run will populate the output folder as follows:

| file/folder | description |
| ------ | ------ |
| complete/       | Folder containing key individual outputs.       |
| inferred_proteins.txt       | Manifest of all input protein sequences iterated through Alphafold.       |
| ss_summary.tsv       | Table of biophysical properties generated for all input proteins.       |
| scatterplot_sizeVssVcharge.pdf | Scatterplot of size, structure, and surface charge for all input proteins. |

Readouts in 'complete' folder currently include:
- Ranked, relaxed, and unrelaxed structure predictions and associated outputs from Alphafold (.pdb, etc).
- Outputs of pdb2pqr, including propka calculation of protein ionization (.txt, .log, .pqr, .in).
- Solvated and vacuum electrostatic maps generated by APBS (.dx, .out).
- Charge vs. pH plots of folded and unfolded states including isoelectric point and charge at neutral pH (.png).


Properties in ss_summary.tsv currently include:
- Protein chain length
- Percentages of alpha helices and beta sheets (also summarized as '%structured')
- Solvation energy (total minus vacuum energy)
 

## To do

- Add additional biophysical parameters.
- Enable support for multi-chain coordinate files.
- Incorporate protein MSA, structural alignment, and phylogenetic analysis modules.
- Automatic generation of electrostatic visualizations from Data Explorer format.
- Add scoring mechanism for biophysical/functional properties likely shared
between multiple proteins.


## Acknowledgements

alphaPhysical references the following packages and libraries:

[Alphafold](https://github.com/google-deepmind/alphafold)\
[APBS](https://www.poissonboltzmann.org/)\
[Anaconda](https://www.anaconda.com/)\
[Docker](https://www.docker.com/)\
[Matplotlib](https://matplotlib.org/)\
[NumPy](https://numpy.org/)\
[Pandas](https://pandas.pydata.org/)\
[PyMOL](https://pymol.org/conda/)\
[Python](https://www.python.org/)


## License and Disclaimer

The MIT License (MIT)

Copyright (c) 2024 Steven Heaton

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


Third-party software

Use of the third-party software, libraries or code referred to in the Acknowledgements section above may be governed by separate terms and conditions or license provisions. Your use of the third-party software, libraries or code is subject to any such terms and you should check that you can comply with any applicable restrictions or terms and conditions before use.
